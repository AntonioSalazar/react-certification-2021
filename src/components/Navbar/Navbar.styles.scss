.navbar {
  display: flex;
  justify-content: space-between;
  &__left-section {
    width: 100%;
    display: flex;
    align-items: center;
    background-color: green;
  }

  &__right-section {
    background-color: orange;
    display: flex;
    align-items: center;
    justify-content: flex-end;
    width: 100%;
    padding: 5px;
  }
}
