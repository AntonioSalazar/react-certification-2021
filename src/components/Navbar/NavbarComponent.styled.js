import styled from 'styled-components';


export const InputText = styled.input`
background-color: var(--main-blue);
border: none;
color: var(--main-white);
margin-left: 2rem;
padding: 0.5rem;
width: 60%;
transition: all 0.3s ease;
&::placeholder {
  color: var(--main-white);
}
&:hover {
  width: 100%;
  border-bottom: 2px solid var(--main-white);
}
`;
InputText.displayName = 'InputText';

export const LogoutBtn = styled.button`
  padding: 1rem;
  border-radius: 4px;
  background-color: var(--main-white);
  color: var(--main-blue);
  transition: all .3s ease;
  border: 1px solid transparent;
  &:hover{
    background-color: var(--main-blue);
    color: var(--main-white);
    border: 1px solid var(--main-white);
  }
`;
LogoutBtn.displayName = 'LogoutBtn'