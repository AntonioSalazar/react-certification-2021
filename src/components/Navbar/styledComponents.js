import styled from 'styled-components';


export const InputText = styled.input`
background-color: #060b26;
border: none;
color: #fff;
margin-left: 2rem;
padding: 0.5rem;
width: 60%;
transition: all 0.3s ease;
&::placeholder {
  color: #fff;
}
&:hover {
  width: 100%;
  border-bottom: 2px solid #fff;
}
`;
InputText.displayName = 'InputText';

export const LogoutBtn = styled.button`
  padding: 1rem;
  border-radius: 4px;
  background-color: #fff;
  color: #060b26;
  transition: all .3s ease;
  border: 1px solid transparent;
  &:hover{
    background-color: #060b26;
    color: #fff;
    border: 1px solid #fff;
  }
`;
LogoutBtn.displayName = 'LogoutBtn'