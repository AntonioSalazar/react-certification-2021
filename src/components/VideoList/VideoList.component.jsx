import React from 'react';
import styled from 'styled-components';

const VideoCardDiv = styled.div`
  display: flex;
  flex-direction: column;
  width: 30rem;
  height: 35rem;
  border-radius: 0.4rem;
  margin: 2rem;
  border: 1px #dad4d4 solid;
  transition: all 0.3s ease;
  &:hover {
    box-shadow: 10px 10px 24px 0px rgba(0, 0, 0, 0.75);
  }
`;

const VideoImageDiv = styled.div`
  height: 140px;
`;

const TitleSpan = styled.span`
  font-size: 1.6rem;
  font-weight: bold;
  margin: 2rem;
`;

const DescriptionSpan = styled.span`
  font-size: 1.4rem;
  margin-bottom: 2rem;
  text-align: left;
  margin: 2rem;
`;

const VideoList = ({ video }) => {
  const { title, description, thumbnails } = video;

  return (
    <VideoCardDiv>
      <VideoImageDiv
        style={{
          backgroundImage: `url(${thumbnails.medium.url})`,
          backgroundSize: 'cover',
          backgroundRepeat: 'no-repeat',
          backgroundPosition: 'center',
          borderTopRightRadius: '0.4rem',
          borderTopLeftRadius: '0.4rem',
        }}
      ></VideoImageDiv>
      <TitleSpan>{title}</TitleSpan>
      <DescriptionSpan>{description ? description : 'watch video...'}</DescriptionSpan>
    </VideoCardDiv>
  );
};

export default VideoList;
