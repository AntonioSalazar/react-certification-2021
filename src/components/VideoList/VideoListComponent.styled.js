import styled from 'styled-components';

export const VideoCardDivLight = styled.div`
  display: flex;
  flex-direction: column;
  width: 30rem;
  height: 35rem;
  border-radius: 0.4rem;
  margin: 2rem;
  border: 1px var(--light-gray) solid;
  transition: all 0.3s ease;
  &:hover {
      box-shadow: 10px 10px 24px 0px var(--black);
    }
    
    @media only screen and (max-width: 670px) {
        width: 80%;
    }
`;
VideoCardDivLight.displayName = 'VideoCardDivLight';

export const VideoCardDivDark= styled.div`
  display: flex;
  flex-direction: column;
  width: 30rem;
  height: 35rem;
  border-radius: 0.4rem;
  margin: 2rem;
  border: 1px var(--light-gray) solid;
  transition: all 0.3s ease;
  &:hover {
      -webkit-box-shadow: 5px 5px 15px 5px var(--black); 
      box-shadow: 5px 5px 15px 5px var(--light-gray) ;
    }
    
    @media only screen and (max-width: 670px) {
        width: 80%;
    }
`;
VideoCardDivDark.displayName = 'VideoCardDivDark';

export const VideoImageDiv = styled.div`
  height: 140px;
`;
VideoImageDiv.displayName = 'VideoImageDiv';

export const TitleSpan = styled.span`
  font-size: 1.6rem;
  font-weight: bold;
  margin: 2rem;
`;
TitleSpan.displayName = 'TitleSpan';

export const DescriptionSpan = styled.span`
  font-size: 1.4rem;
  margin-bottom: 2rem;
  text-align: left;
  margin: 2rem;
  `;
DescriptionSpan.displayName = 'DescriptionSpan';

