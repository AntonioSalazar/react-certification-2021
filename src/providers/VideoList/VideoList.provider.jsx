import React, { createContext, useEffect, useState } from 'react';
export const VideoListContext = createContext();
const VideoListProvider = (props) => {
  const [videos, setVideos] = useState([]);
  const [search, setSearch] = useState('wizeline');

  useEffect(() => {
    const getVideos = async () => {
      try {
        const url = `https://www.googleapis.com/youtube/v3/search?part=snippet&&key=${process.env.REACT_APP_GOOGLE_APP_API_KEY}&type=video&maxResults=23&q=${search}`;
        const response = await fetch(url);
        const data = await response.json();
        setVideos(data.items);
      } catch (error) {
        throw new Error(error);
      }
    };
    getVideos();
  }, [search]);

  return (
    <VideoListContext.Provider
      value={{
        setSearch,
        videos,
      }}
    >
      {props.children}
    </VideoListContext.Provider>
  );
};

export default VideoListProvider;
