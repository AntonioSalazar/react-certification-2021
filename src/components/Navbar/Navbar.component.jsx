import React, { useState, useContext } from 'react';
import { Link } from 'react-router-dom';
import { FaHamburger, FaWindowClose, FaHome, FaUserNinja } from 'react-icons/fa';
import { IconContext } from 'react-icons/lib';
import './Navbar.styles.scss';
import styled from 'styled-components';
import { VideoListContext } from '../../providers/VideoList/VideoList.provider';

const Input = styled.input`
  background-color: #060b26;
  border: none;
  color: #fff;
  margin-left: 2rem;
  padding: 0.5rem;
  width: 60%;
  transition: all 0.3s ease;
  &::placeholder {
    color: #fff;
  }
  &:hover {
    width: 100%;
    border-bottom: 2px solid #fff;
  }
`;

Input.displayName = 'Input';

// ##### Im thinking about refactoring this component into another component as I think it's kind of complex to read ####

const Navbar = () => {
  const [sidebar, setSidebar] = useState(false);
  const [inputValue, setInputValue] = useState('');
  const toogleSideBar = () => setSidebar(!sidebar);
  const { setSearch } = useContext(VideoListContext);

  const handleSubmit = (e) => {
    e.preventDefault();
    setSearch(inputValue);
  };

  return (
    <>
      <IconContext.Provider value={{ color: '#fff', size: '3rem' }}>
        <nav className="navbar">
          <div className="navbar__left-section">
            <Link to="/" className="menu-bars">
              <FaHamburger onClick={toogleSideBar} />
            </Link>
            <form onSubmit={handleSubmit}>
              <Input
                type="text"
                name="search"
                placeholder="search a video..."
                value={inputValue}
                onChange={(e) => setInputValue(e.target.value)}
              />
            </form>
          </div>

          <div className="navbar__right-section">
            <div className="toggle">
              <label className="switch">
                <input type="checkbox" />
                <span className="slider round"></span>
              </label>
            </div>
            <Link to="/">
              <FaUserNinja />
            </Link>
          </div>
        </nav>

        <nav className={sidebar ? 'nav-menu active' : 'nav-menu'}>
          <ul className="nav-menu-items" onClick={toogleSideBar}>
            <li className="navbar-toggle">
              <Link to="/" className="menu-bars">
                <FaWindowClose />
              </Link>
            </li>
            <li className="nav-text">
              <Link to="/">
                <FaHome />
                <span>Home</span>
              </Link>
            </li>
          </ul>
        </nav>
      </IconContext.Provider>
    </>
  );
};

export default Navbar;
