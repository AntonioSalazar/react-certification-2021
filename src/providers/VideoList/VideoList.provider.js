import React, { createContext, useEffect, useState, useReducer } from 'react';
import { useHistory } from "react-router-dom";
import { reducer } from './VideoList.reducer';
import Swal from 'sweetalert2'
export const VideoListContext = createContext();

const VideoListProvider = (props) => {
  const [search, setSearch] = useState('wizeline');
  const initialState = {
    loading: true,
    posts: [],
    error: '',
    favoriteVideos: []
  };
  const [videosState, dispatch] = useReducer(reducer, initialState);
  const history = useHistory()
  useEffect(() => {
    let mounted  = true;
    const getVideos = async () => {
        try {
          const url = `https://www.googleapis.com/youtube/v3/search?part=snippet&&key=${process.env.REACT_APP_GOOGLE_APP_API_KEY}&type=video&maxResults=23&q=${search}`;
          const response = await fetch(url);
          if (mounted) {
            const data = await response.json();
            dispatch({ type: 'FETCH_SUCCESS', payload: data.items });
          }
          history.push('/')
        } catch (error) {
          dispatch({ type: 'FETCH_ERROR' });
          throw new Error(error);
        }
      
    };
    getVideos();
    return () => mounted = false;

  }, [search, history]);

  const addToFavorites = (video) => {
    dispatch({
      type: 'ADD_TO_FAVORITES',
      payload: video
    })
    Swal.fire(
      'Video added to your favorites'
    )
  };

  const removeFromFavorites = id => {
    dispatch({
      type: 'REMOVE_FROM_FAVORITES',
      payload: id
    })
    Swal.fire(
      'Video has been removed from favorites'
    )
  }

  return (
    <VideoListContext.Provider
      value={{
        setSearch,
        addToFavorites,
        videosState, 
        removeFromFavorites
      }}
    >
      {props.children}
    </VideoListContext.Provider>
  );
};

export default VideoListProvider;
