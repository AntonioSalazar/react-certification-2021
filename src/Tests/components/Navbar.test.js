import React from 'react';
import { mount } from 'enzyme';
import Navbar from '../../components/Navbar/Navbar.component'
import { VideoListContext } from '../../providers/VideoList/VideoList.provider';
import Context from '../../providers/Theme/Theme.provider'
import { MemoryRouter } from 'react-router';

describe('Tests of the Navbar component', () => {
    
    const setSearch = jest.fn()
    const dispatch = jest.fn()
    const handleClick = jest.fn()
    let wrapper;
    beforeEach(() =>  {
        jest.clearAllMocks();
        wrapper = mount(
            <Context.Provider
                value={{
                    dispatch
                }}
            >
                <VideoListContext.Provider
                    value={{
                        setSearch
                    }}
                    >
                    <MemoryRouter>
                        <Navbar/>
                    </MemoryRouter>
                </VideoListContext.Provider>
            </Context.Provider>)
    })

    test('should render the Navbar component correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })
    
    test('should have an styled input tag', () => {
        const styledInput = wrapper.find('InputText')
        const value = 'React!'
        styledInput.simulate('change', {
            target: {
                value
            }
        })
        expect(styledInput.length).toBe(1)
    })
    test('should post info on submit', () => {
        const formEventMocked = { preventDefault: setSearch };
        wrapper.find('form').simulate('submit', formEventMocked);       
        expect(wrapper.find('form')).toHaveLength(1);
        expect(formEventMocked.preventDefault).toHaveBeenCalledTimes(1)
    })
    
    test('should call setSearch and clean the input text ', () => {
        const value = 'learn jest'
        wrapper.find('InputText').simulate('change', { target: {value}})
        wrapper.find('form').simulate('submit', {preventDefault(){}});
        expect(setSearch).toHaveBeenCalled()
        expect(wrapper.find('InputText').prop('value')).toBe('')

    })
    
    
})

