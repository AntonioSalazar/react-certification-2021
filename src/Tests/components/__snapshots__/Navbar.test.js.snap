// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of the Navbar component should render the Navbar component correctly 1`] = `undefined`;
